### ISSUE TITLE
Cache DNS Lookup 

### ISSUE THREAD

**Issue by arianvp** [channable/vaultenv#18]
Some of our components require hundreds of secrets to be collected from vault at once.   (We have hundreds of services requesting hundreds of secrets in bursts sometimes)  This was actually hitting limits of our DNS resolver.   Simply doing the lookup once, and then use it for consequent http requests would solve this.

**Comment by arianvp** [comment]
I think we should solve this at the OS level.    There is probably a caching parameter that we're missing.

However,  I think we could reduce the amount of requests that we send for a secrets file by only requesting each path once.    I'll file a new issue for that.

**Comment by rkrzr** [comment]
> However, I think we could reduce the amount of requests that we send for a secrets file by only requesting each path once.

Do you mean for each "store" (like e.g. `secret/sentry`)? That's a good idea.

**Comment by arianvp** [comment]
@rkrzr yes!  exactly.

**Comment by duijf** [comment]
In our production setup, we currently cache DNS lookups on the host level. This is IMO the better place for this: vaultenv shouldn't attempt to cache this. Otherwise it would explicitly override the TTLs for the `vault.service.consul` DNS record, which I'd call unexpected behavior.

For any other people running into this: you can configure consul to set a TTL on the vault service definition. The rest should be handled automatically. Config snippet:

```json
{
  "dns_config": {
    "service_ttl": {
      "vault": "5s"
    }
  }
}
```

More info here: https://www.consul.io/docs/guides/dns-cache.html

**Comment by rkrzr** [comment]
> In our production setup, we currently cache DNS lookups on the host level.

The title of this issue is a bit misleading. We still don't cache any DNS lookups, what we instead do, is get every secret (which can contain multiple key-value pairs) only once, instead of for every key.

The associated PR is clearer: https://github.com/channable/vaultenv/pull/27

**Comment by duijf** [comment]
> We still don't cache any DNS lookups, what we instead do, is get every secret (which can contain multiple key-value pairs) only once, instead of for every key.

Both statements are true. We do the DNS caching, but dnsmasq takes care of it, not vaultenv :smile:
